NetKet/Supervised/supervised.hpp

```diff
@@ -0,0 +1,119 @@
+// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_SUPERVISED_CC
+#define NETKET_SUPERVISED_CC
+
+#include <memory>
+#include <string>
+
+#include "Optimizer/optimizer.hpp"
+#include "vmc.hpp"
+
+namespace netket {
+
+class Supervised {
+  using VectorType = Eigen::Matrix<std::complex<double>, Eigen::Dynamic, 1>;
+
+  int batchsize_ = 100;
+
+  Eigen::MatrixXd inputs_;
+  Eigen::VectorXcd targets_;
+
+ public:
+  explicit Supervised(const json &supervised_pars) {
+    // Relevant parameters for supervised learning
+    // is stored in supervised_pars.
+    CheckFieldExists(supervised_pars, "Supervised");
+    const std::string loss_name =
+        FieldVal(supervised_pars["Supervised"], "Loss", "Supervised");
+
+    // Input data is encoded in Json file with name "InputFilename".
+    auto data_json =
+        ReadJsonFromFile(supervised_pars["Supervised"]["InputFilename"]);
+    using DataType = Data<double>;
+    DataType data(data_json, supervised_pars);
+
+    // Make a machine using the Hilbert space extracted from
+    // the data.
+    using MachineType = Machine<std::complex<double>>;
+    MachineType machine(data.GetHilbert(), supervised_pars);
+
+    if (loss_name == "Overlap") {
+      // To do:
```

Comment:
The Overlap loss function seems the most problematic to me, also code-wise. 
The MSE implementation is very nice and compact. Wouldn't it be possible to have something very similar also for the Overlap loss? Since we don't need sampling, it should be possible to have a very similar function. Even better option would be to have a supervised class worker templatized on a generic loss function. Basically one would like to have a single piece of code that works with different losses, provided that the loss is an object returning its gradient

Reply:
Indeed, for the new formula we discussed, one works with uniform sampling of data given. (I guess this is what you mean by no sampling? ) In this case, there is no need to have the duplicated VMC for the calculating the overlap loss optimization. In principle, we could handle  this compactly in supervised.hpp file.

Unfortunately, the new formula did not give converge results so far. Maybe it is because of the test case chosen. I will test a few more example. But before that, the working version of overlap loss optimization contains the VMC part.